src/actions/channel_categories.ts

```diff
@@ -16,3 +48,343 @@ export function collapseCategory(categoryId: string) {
+                    ...category,
+                    channel_ids: insertWithoutDuplicates(category.channel_ids, channel.id, 0),
+                })),
+            });
+        }
+
+        // Add the new channel to the Channels category on the channel's team
+        if (categories.some((category) => category.channel_ids.some((channelId) => channelId === channel.id))) {
+            return {data: false};
+        }
+        const channelsCategory = getCategoryInTeamByType(state, channel.team_id, CategoryTypes.CHANNELS);
+
+        if (!channelsCategory) {
+            // No categories were found for this team, so the categories for this team haven't been loaded yet.
+            // The channel will have been added to the category by the server, so we'll get it once the categories
+            // are actually loaded.
+            return {data: false};
+        }
+
+        if (setOnServer) {
+            return dispatch(addChannelToCategory(channelsCategory.id, channel.id));
+        }
+
+        return dispatch({
+            type: ChannelCategoryTypes.RECEIVED_CATEGORY,
+            data: {
+                ...channelsCategory,
+                channel_ids: insertWithoutDuplicates(channelsCategory.channel_ids, channel.id, 0),
+            },
+        });
+    };
+}
+
+// addChannelToCategory returns an action that can be dispatched to add a channel to a given category without specifying
+// its order. The channel will be removed from its previous category (if any) on the given category's team and it will be
+// placed first in its new category.
+export function addChannelToCategory(categoryId: string, channelId: string): ActionFunc {
+    return moveChannelToCategory(categoryId, channelId, 0, false);
+}
+
+// moveChannelToCategory returns an action that moves a channel into a category and puts it at the given index at the
+// category. The channel will also be removed from its previous category (if any) on that category's team. The category's
+// order will also be set to manual by default.
+export function moveChannelToCategory(categoryId: string, channelId: string, newIndex: number, setManualSorting = true) {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const state = getState();
+        const category = getCategory(state, categoryId);
+        const currentUserId = getCurrentUserId(state);
+
+        // Add the channel to the new category
+        const categories = [{
+            ...category,
+            sorting: (setManualSorting && category.type !== CategoryTypes.DIRECT_MESSAGES) ? CategorySorting.Manual : category.sorting,
+            channel_ids: insertWithoutDuplicates(category.channel_ids, channelId, newIndex),
+        }];
+
+        // And remove it from the old category
+        const originalCategory = getCategoryInTeamWithChannel(getState(), category.team_id, channelId);
+        if (originalCategory && originalCategory.id !== category.id) {
```

Comment:
why would we move to the same category? is it to change it's index?

Reply:
Yep, that's correct

Reply:
Rearranging?